bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -1916,6 +1924,33 @@ private static boolean checkArrayEquivalent(Type actualType, Type expType) {
         return expType == actualType;
     }
 
+    private static boolean checkIsNeverTypeOrStructureTypeWithARequiredNeverMember(Type type) {
+        switch (type.getTag()) {
+            case TypeTags.NEVER_TAG:
+                return true;
+            case TypeTags.RECORD_TYPE_TAG:
+                for (Field field : ((BRecordType) type).getFields().values()) {
+                    // skip check for fields with self referencing type and not required fields.
+                    if (SymbolFlags.isFlagOn(field.getFlags(), SymbolFlags.REQUIRED) &&
+                            !isSameType(type, field.getFieldType()) &&
+                            checkIsNeverTypeOrStructureTypeWithARequiredNeverMember(field.getFieldType())) {
+                        return true;
+                    }
+                }
+                return false;
+            case TypeTags.TUPLE_TAG:
```

Comment:
Shouldn't we consider a closed array of type `never` to be `never`? e.g., `never[3]`?